composer/utils/object_store/mlflow_object_store.py

```diff
@@ -0,0 +1,392 @@
+            the currently active MLFlow run if one exists. If no active run exists, a new run will be created under a
+            default experiment name, or the experiment name specified by the `MLFLOW_EXPERIMENT_NAME` environment
+            variable if one is set.
+
+            If `experiment_id` is provided and `run_id` is not, the MLFlowObjectStore will create a new run in the
+            provided experiment.
+
+            Providing a `run_id` without an `experiment_id` will raise an error.
+        multipart_upload_chunk_size(int, optional): The maximum size of single chunk in an MLFlow multipart upload.
+            The maximum number of chunks supported by MLFlow is 10,000, so the max file size that can
+            be uploaded is `10 000 * multipart_upload_chunk_size`. Defaults to 100MB for a max upload size of 1TB.
+    """
+
+    def __init__(self, path: str, multipart_upload_chunk_size: int = 100 * 1024 * 1024) -> None:
+        try:
+            import mlflow
+            from mlflow import MlflowClient
+        except ImportError as e:
+            raise MissingConditionalImportError('mlflow', conda_package='mlflow>=2.9.2,<3.0') from e
+
+        try:
+            from databricks.sdk import WorkspaceClient
+        except ImportError as e:
+            raise MissingConditionalImportError('databricks', conda_package='databricks-sdk>=0.15.0,<1.0') from e
+
+        tracking_uri = os.getenv('MLFLOW_TRACKING_URI', MLFLOW_DATABRICKS_TRACKING_URI)
+        if tracking_uri != MLFLOW_DATABRICKS_TRACKING_URI:
+            raise ValueError(
+                'MLFlowObjectStore currently only supports Databricks-hosted MLFlow tracking. '
+                f'Environment variable `MLFLOW_TRACKING_URI` is set to a non-Databricks URI {tracking_uri}. '
+                f'Please unset it or set the value to `{MLFLOW_DATABRICKS_TRACKING_URI}`.')
+
+        # Use the Databricks WorkspaceClient to check that credentials are set up correctly.
+        try:
+            WorkspaceClient()
+        except Exception as e:
+            raise ValueError(
+                f'Databricks SDK credentials not correctly setup. '
+                'Visit https://databricks-sdk-py.readthedocs.io/en/latest/authentication.html#databricks-native-authentication '
+                'to identify different ways to setup credentials.') from e
+
+        self._mlflow_client = MlflowClient(MLFLOW_DATABRICKS_TRACKING_URI)
+        mlflow.environment_variables.MLFLOW_MULTIPART_UPLOAD_CHUNK_SIZE.set(multipart_upload_chunk_size)
+
+        experiment_id, run_id, _ = MLFlowObjectStore.parse_dbfs_path(path)
+        if experiment_id == PLACEHOLDER_EXPERIMENT_ID:
+            experiment_id = None
+        if run_id == PLACEHOLDER_RUN_ID:
+            run_id = None
+
+        # Construct the `experiment_id` and `run_id` depending on whether format placeholders were provided.
+        if not dist.is_initialized() or dist.get_global_rank() == 0:
```

Comment:
do you need the `dist.is_initialized` check for some reason?

Reply:
It was necessary for manual testing (i.e. just initializing an `MLflowObjectStore` directly). Without it, it would error out on the `dist` functions. More generally, I don't think the object store should necessarily assume it's being used in a distributed context from a modular design perspective.

Reply:
the dist functions are all supposed to be usable in a non dist environment, so nothing should be erroring out. e.g. see the logic here https://github.com/mosaicml/composer/blob/2aa50e7741a077ff21f5743934fbcf4b755d441e/composer/utils/dist.py#L346-L376

Reply:
Hmm let me try it again without the check. I was definitely running into a problem that adding the check seemed to solve, but maybe it was symptomatic of a different issue since this was when we were still playing around with all the different ways to broadcast the experiment/run IDs

Reply:
Looks like it works without the check 🤔. Will remove and verify in the follow-up PR that everything looks good E2E

Reply:
Actually just remembered why I needed to add the checks - the object store initialization fails in the RUD workers without it since it's running in a different process:
```
Process SpawnProcess-1:
Traceback (most recent call last):6-ac8f-e18cce030076/ift-mpt-7b-vymnpr/checkpoints/ep4-ba204-rank0.pt:   0%|          | 0.00/49.5G [00:00<?, ?iB/s]
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3/dist-packages/composer/loggers/remote_uploader_downloader.py", line 624, in _upload_worker
    remote_backend = _build_remote_backend(remote_backend_name, backend_kwargs)
  File "/usr/lib/python3/dist-packages/composer/loggers/remote_uploader_downloader.py", line 69, in _build_remote_backend
    return remote_backend_cls(**backend_kwargs)
  File "/usr/lib/python3/dist-packages/composer/utils/object_store/mlflow_object_store.py", line 147, in __init__
    dist.broadcast_object_list(mlflow_info, src=0)
  File "/usr/lib/python3/dist-packages/composer/utils/dist.py", line 371, in broadcast_object_list
    raise RuntimeError(f'The world_size({world_size}) > 1, but the distributed package is not '
RuntimeError: The world_size(8) > 1, but the distributed package is not available or has not been initialized. Please check you have initialized the distributed runtime and that PyTorch has been built with distributed support. If calling this function outside Trainer, please ensure that `composer.utils.dist.initialize_dist` has been called first.
```

Reply:
ah right. i think that just means the init isn't happening the way you intend it to? That is, the object store constructor is called in the RUD subprocesses, so dist won't be initialized. Didn't we discuss putting the dist sync logic in RUD instead of in the object store constructor itself? i might be misremembering.

Reply:
Right - for the RUD workers, I broadcast the updated path in the RUD https://github.com/mosaicml/composer/blob/02a373343c7a0188f2f88b289320b31cf901f550/composer/loggers/remote_uploader_downloader.py#L344-L356

But we still need to do a sync in the object store constructor for cases where the object store is used outside a RUD (e.g., `maybe_create_object_store_from_uri()`). We need some way to gate which instances actually attempt to initialize, which would typically just be rank 0, but that will fail in the workers. I suppose I could pull out the case where both IDs are provided to avoid any sort of `dist` operations in that scenario, but that seems uglier/less readable than just checking for `dist` initialization.
```
if experiment_id is not None and run_id is not None:
  run = self._mlflow_client.get_run(run_id)
  if run.info.experiment_id != experiment_id:
      raise ValueError(
          f'Provided `run_id` {run_id} does not belong to provided experiment {experiment_id}. '
          f'Found experiment {run.info.experiment_id}.')
else:
  if dist.get_global_rank() == 0:
    experiment_id, run_id = init_run_info()

  mlflow_info = [experiment_id, run_id]
  dist.broadcast_object_list(mlflow_info, src=0)
  experiment_id, run_id = mlflow_info
  

```

It also restricts how the object store can be used in subprocesses by imposing the requirement that the IDs must be provided, while checking `dist.is_initialized()` doesn't.

Reply:
Ah I see. I think I had been thinking we would force `MLFlowObjectStore` construction to always have the experiment and run id. For most of our cases, RUD would handle populating these. Do we have a use case for `MLFlowObjectStore` needing to create stuff on its own or can we just error with it being not supported? I'm also open to keeping your logic, just prefer to avoid extra distributed logic wherever possible.

Reply:
We need it for autoresume, which creates the object store with `maybe_create_object_store_from_uri` instead of `maybe_create_remote_uploader_downloader_from_uri`. In particular, the returned object store in the first resumption won't have the IDs populated in its path, so it needs to run the init logic that checks the active run.

Reply:
Thanks for the explanation! Sorry I have forgotten some of our past conversations 😅 What do you think about adding the same sync logic to `maybe_create_object_store_from_uri` instead of putting it in the object store init?